Fix #7663: 3.20: Free Plan – Adding (up to 3) URLs from Global Score widget when no credits left isn’t working

Miraeld opened the pull request:
# Description

Fixes #7663
Allow the user to add URLs up to the maximum limit even tho he doesn't have any credit.


## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).

## Detailed scenario

### What was tested

Adding URL when no credit, 
Disabled button when max limit is reached.

### How to test

Refer to issue.

## Technical description

### Documentation

This pull request implements a user plan-based URL limit for the Performance Monitoring feature, ensuring users cannot add more URLs than their license allows. The changes span backend logic, AJAX controllers, frontend rendering, and JavaScript, providing real-time feedback and disabling relevant UI elements when the limit is reached.

**Backend logic and API integration:**

* Added a method to calculate the remaining number of URLs a user can add, based on their plan limit, in both the `Controller` and AJAX controller classes (`get_remaining_url_count`). This is used throughout the backend to enforce the limit. [[1]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R359-R378) [[2]](diffhunk://#diff-f0deed186b1875e34b4fb1dced4935e7d95379016d9394dbff6b39c4e7ac94e5R309-R320)
* Injected the `User` API client into the relevant controllers to access plan and limit data. [[1]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R14) [[2]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R55-R61) [[3]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R70-R78) [[4]](diffhunk://#diff-77d1eb49c422a18485050ab6cb8444dca5e887f0094c35c0b865a7c31f5c690bR149)

**AJAX and data payload updates:**

* Updated AJAX responses (e.g., when adding or resetting a page) to include the `remaining_urls` value, ensuring the frontend always knows the current limit state. [[1]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R143) [[2]](diffhunk://#diff-26007bb022d8ac7e2f984a1965468bf851a0a5c01ab4624694ada073e73250d9R338) [[3]](diffhunk://#diff-193ec9907d4056443da3d2b7b8b1d201a9519f1c801e00b9f5649cddb6657772R165) [[4]](diffhunk://#diff-193ec9907d4056443da3d2b7b8b1d201a9519f1c801e00b9f5649cddb6657772L246-R249) [[5]](diffhunk://#diff-193ec9907d4056443da3d2b7b8b1d201a9519f1c801e00b9f5649cddb6657772R271)

**Frontend rendering and UI feedback:**

* Modified rendering logic for the global score widget and URLs table to include flags (`can_add_url`, `reach_max_url`) for disabling/enabling the "Add URL" button and showing appropriate tooltips when the limit is reached. [[1]](diffhunk://#diff-f418fc954a081f3c507fbe3e5b00010e2562e99598a1c5dc2a8a939e9409c886R75-R77) [[2]](diffhunk://#diff-f418fc954a081f3c507fbe3e5b00010e2562e99598a1c5dc2a8a939e9409c886R112-R114) [[3]](diffhunk://#diff-10b0d27bad50740a30ec95c09e7dae3645cdc0c031a02cd8c59b5107487fbc9aL46-R49) [[4]](diffhunk://#diff-9949bcd4678bb9862456cccf635c40519ed0372579fcd54d544c9cd311ba53b7L106-R108)

**JavaScript enhancements:**

* Added logic to disable or enable the "Add URL" button and input fields in real time, based on the current `remaining_urls` value from AJAX responses or initial page load. Tooltips are updated accordingly to inform users when the maximum is reached. [[1]](diffhunk://#diff-043c510ef3475c349d06d41dcc8b34e2c509f6eda990d9de5a1686b06e46126cL16-R22) [[2]](diffhunk://#diff-043c510ef3475c349d06d41dcc8b34e2c509f6eda990d9de5a1686b06e46126cR308-R345) [[3]](diffhunk://#diff-043c510ef3475c349d06d41dcc8b34e2c509f6eda990d9de5a1686b06e46126cR464-R468) [[4]](diffhunk://#diff-043c510ef3475c349d06d41dcc8b34e2c509f6eda990d9de5a1686b06e46126cR505-R507) [[5]](diffhunk://#diff-043c510ef3475c349d06d41dcc8b34e2c509f6eda990d9de5a1686b06e46126cR524-R526)

These changes collectively ensure that users cannot exceed their URL quota, and the UI provides immediate, clear feedback when the limit is reached.

### New dependencies

None

### Risks

None

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification

No test for that

Miraeld committed:
Enhance performance monitoring: update URL addition logic and tooltips based on credit status

Miraeld committed:
feat: Add dynamic URL limit state management to Performance Monitoring UI

Miraeld committed:
Move method from sub to controller

Miraeld committed:
Refactor prototype

Miraeld committed:
fix: Remove unnecessary comments in Render class methods

Miraeld committed:
Fix test

wordpressfan commented on src/js/global/ajax.js:
```diff
@@ -299,6 +305,44 @@ document.addEventListener('DOMContentLoaded', function() {
+		
+		$('.wpr-pma-global-score-add-url-button').addClass('disabled').attr('disabled', true);
+		
+		const tooltipMessage = 'Maximum number of URLs reached for your license.';
```
can we pass a translated message from php to js?

Miraeld replied:
Done

wordpressfan reviewed (changes requested):
Many thanks @Miraeld 
I have two questions here:
1. Why do we check remaining_urls with retest?
2. Here we handle disabling the button with JS, even for the first load of the page but we also handle it in the view so do we really need this double-work?

Thanks again, that's amazing.

Miraeld committed:
Remove JS init button

Miraeld committed:
Adds translation to error message

Miraeld committed:
Remove useless check while retesting

Miraeld committed:
Add tests

Miraeld committed:
Reset baseline to ignore tests files

Miraeld commented:
@wordpressfan 
1. Why do we check remaining_urls with retest?
I added it as a possible security we could use later, but well it's true, it's not needed. So to avoid adding useless things and complexity, I've removed it. 

2. Here we handle disabling the button with JS, even for the first load of the page but we also handle it in the view so do we really need this double-work?
I've removed the JS part for the init load.

remyperona committed:
fix test

wordpressfan committed:
Merge branch 'feature/3.20' into fix/7663-add-page-when-no-credit
